File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        
        deque<ListNode*> window = {head};
        
        
        while (window.size() < n){
            window.push_back(window.back()->next);
        }
        
        if (window.back()->next == nullptr){
            
            if (window.size()>=2) {return window[1];}              
            else{
                return nullptr;
            }
                
        }
        
        window.push_back(window.back()->next);
        while (window.back()->next != nullptr){
            window.pop_front();
            window.push_back(window.back()->next);
        }
        
        if (window.size()>=3){
            window[0]->next = window[2];
        }
        else{
            window[0]->next = nullptr;
        }
        
        return head;
    }
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        
        queue = [head]
        
        while len(queue) < n:
            queue.append(queue[-1].next)
        
        if queue[-1].next is None:
            try:
                return queue[1]
            except:
                return None
        
        queue.append(queue[-1].next)
        while queue[-1].next is not None:
            queue.pop(0)
            queue.append(queue[-1].next)
            
        try:
            queue[0].next = queue[2]
        except:
            queue[0].next = None
            
        return head
        